### PR TITLE
Migration Spark 2

### PR DESCRIPTION
**Pull request by NastasiaSaby**

### CODE REVIEW

File: _posts/2017-12-13-spark-2.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+description: "Feedback from a migration to Spark 2.2"
+author:
+  name: Nastasia Saby
+  avatar:
+  email:
+  twitter: saby_nastasia
+  facebook:
+  github: 
+category:
+tags: [Data,Hadoop,BigData,Airflow,Hive,Spark]
```
Comment (LaurentVion) [inline _posts/2017-12-13-spark-2.md]: java

File: _posts/2017-12-13-spark-2.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  credit: 
+  creditlink: 
+comments: true
+language: en
+---
+
+For big data needs, we use Spark. We recently migrated our biggest project from Spark 1.5 to Spark 2.2 and wanted to tell that story.
+
+Spark 2 has been released a year ago (July 26, 2016). So, maybe we are a bit late. But, better late than never.
+
```
Comment (LaurentVion) [inline _posts/2017-12-13-spark-2.md]: Ca serait bien de mettre une phrase sur la partie cloudera.

File: _posts/2017-12-13-spark-2.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+Then, we ran our jobs with some data sets. The idea was to check the differences with Spark 1.5.
+
+We wanted to be sure that our unit parts were working together.
+
+## d) Double run
+
+Then, we set a double run. It means that we had our jobs running with Spark 1.5 and our jobs running with Spark 2.2 and we checked the differences between the two.
+
+We used Airflow to deal with that. So, if you know Airflow, you will understand that we added a new DAG to run our project with Spark 2.2.
```
Comment (LaurentVion) [inline _posts/2017-12-13-spark-2.md]: lien vers homepage airflow

File: _posts/2017-12-13-spark-2.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+tags: [Data,Hadoop,BigData,Airflow,Hive,Spark]
+image:
+  feature: 
+  credit: 
+  creditlink: 
+comments: true
+language: en
+---
+
+For big data needs, we use Spark. We recently migrated our biggest project from Spark 1.5 to Spark 2.2 and wanted to tell that story.
```
Comment (LaurentVion) [inline _posts/2017-12-13-spark-2.md]: lien homepage spark

File: _posts/2017-12-13-spark-2.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+category:
+tags: [Data,Hadoop,BigData,Airflow,Hive,Spark]
+image:
+  feature: 
+  credit: 
+  creditlink: 
+comments: true
+language: en
+---
+
```
Comment (LaurentVion) [inline _posts/2017-12-13-spark-2.md]: Petite intro sur le programme data (Cluster, etc ...). Un lien vers l'article de Cory peut suffire

File: _posts/2017-12-13-spark-2.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+We decided not to use "SparkSession" in a first step and to follow the error instructions. 
+
+So, we replaced our direct call to Hive by a UDF registration.
+
+## e) Depreciations
+
+We noticed some depreciations like HiveContext or Accumulators for instance. But we decided not to deal with them for the moment.
+
+## d) Performance
+We have made some gains in performance. We hope we will make some other gains by migrating to the Spark 2.2 philosophy ("SparkSession", etc).
```
Comment (LaurentVion) [inline _posts/2017-12-13-spark-2.md]: Chiffres

File: _posts/2017-12-13-spark-2.md (hunk 5 above)
Comment (NastasiaSaby) [reply _posts/2017-12-13-spark-2.md]: Mis ici : You can find more information in a [precedent article about our datalake](http://tech.m6web.fr/data/genesis-of-m6-datalake/)

Comment (malikba) [reply _posts/2017-12-13-spark-2.md]: You can find more information in a __previous__ article about our datalake 😄

File: _posts/2017-12-13-spark-2.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  feature: 
+  credit: 
+  creditlink: 
+comments: true
+language: en
+---
+
+For big data needs, we use [Spark](https://spark.apache.org/). You can find more information in a [precedent article about our datalake](http://tech.m6web.fr/data/genesis-of-m6-datalake/). We recently migrated our biggest project from Spark 1.5 to Spark 2.2 and wanted to tell that story.
+
+Spark 2 has been released a year ago (July 26, 2016). So, maybe we are a bit late. But, better late than never.
```
Comment (malikba) [inline _posts/2017-12-13-spark-2.md]: `So, maybe we are a bit late, but better late than never.`

File: _posts/2017-12-13-spark-2.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+We are working with an official version from [Cloudera](https://www.cloudera.com/) with Spark 1.6 as the default version.
+
+Our project runs everyday to get data from different sources and send them to different destinations.
+
+It is built with Java with Spark 1.5. We met several problems with that. First of all, as the community for Java is smaller than for Scala or Python in the Spark world, the community for Spark 1.5 is smaller than for Spark 2.2.
```
Comment (malikba) [inline _posts/2017-12-13-spark-2.md]: `It is built with Java __and__ Spark 1.5, but we met several problems with those technologies. First of all, the Java community is smaller than the ones of Python or Scala. Secondly, the Spark 1.5 community is also smaller than the one of version 2.2.`

File: _posts/2017-12-13-spark-2.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+So, sometimes we got difficulties to find information
+But most of all, we did not success to integrate new components that work with newer versions.
```
Comment (malikba) [inline _posts/2017-12-13-spark-2.md]: `we did not succeed`

File: _posts/2017-12-13-spark-2.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+We wanted to be sure that our unit parts were working together.
+
+## d) Double run
+
+Then, we set a double run. It means that we had our jobs running with Spark 1.5 and our jobs running with Spark 2.2 and we checked the differences between the two.
+
+We used [Airflow](https://airflow.apache.org/) to deal with that. So, if you know Airflow, you will understand that we added a new DAG to run our project with Spark 2.2.
+
+The idea was to see daily the potential differences between the two.
```
Comment (malikba) [inline _posts/2017-12-13-spark-2.md]: The idea was to see the potential differences between the two in a daily basis.

File: _posts/2017-12-13-spark-2.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+We wanted to migrate for fixed bugs in general and in a performance purpose too.
+
+# I) Process
+
+![Process](/images/posts/spark-2/process.png)
+
+## a) Spark 1.5 to Spark 1.6
+
+First, we had decided to migrate to 1.6 to do a progressive migration. But we found a bug with a UDF. We had difficulties to fix it and this bug was no more in 2.2.
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: was no more => was resolved

File: _posts/2017-12-13-spark-2.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+So we finally decided to migrate directly to 2.2.
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: `So` is very oral. try to replace it with something else
here: `In the end, we decided to migrate directly to 2.2.`

File: _posts/2017-12-13-spark-2.md (hunk 10 above)
Comment (Osirisxxl) [reply _posts/2017-12-13-spark-2.md]: **on** a  daily basis

File: _posts/2017-12-13-spark-2.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+## d) Double run
+
+Then, we set a double run. It means that we had our jobs running with Spark 1.5 and our jobs running with Spark 2.2 and we checked the differences between the two.
+
+We used [Airflow](https://airflow.apache.org/) to deal with that. So, if you know Airflow, you will understand that we added a new DAG to run our project with Spark 2.2.
+
+The idea was to see daily the potential differences between the two.
+
+At the end, we merged our branch to the master.
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: into master (remove the)

File: _posts/2017-12-13-spark-2.md [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+Of course, the main change is that "dataFrame" does not exist anymore. You must replace it by "Dataset\<Row\>". 
+
+Actually, "DataFrame" and "Dataset" were unified with Spark 2.0. In reality, for untyped API like Python, "DataFrame" still exists. But, we work with Java.
+
+![Dataset](/images/posts/spark-2/dataset.png)
+
+Using "Dataset\<T\>" is a way to apply a schema at the compilation. So, if there is a problem, you will get a logical exception. Before, with "DataFrame", you could only have runtime exceptions.
+
+So, in a first step, we replaced "DataFrame" by "Dataset\<Row\>"
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: As a first step, we replaced...

File: _posts/2017-12-13-spark-2.md [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+## b) SparkSession
+
+A second major difference is “SparkSession”. It is the new entry to Spark. 
+There is no need anymore to create a "SparkConf", a "SparkContext" and a "SQLContext". It is possible to get all of it only with a "SparkSession".
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: only => just

File: _posts/2017-12-13-spark-2.md [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+But, it is important to understand that if you just want to migrate your code in a first step to get it work with Spark 2, it is not a need to use "SparkSession". "SparkConf", "SparkContext" and "SQLContext" still work.
+
+That is what we decided to do.
+
+## c) Iterable to Iterator
+
+The return type “Iterable” is incompatible with "PairFlatMapFunction". So, we had to replace "Iterable\<\>" with "Iterator\<\>".
+
+We changed code like that:
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: changed => replaced

File: _posts/2017-12-13-spark-2.md [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+java.lang.UnsupportedOperationException: Use sqlContext.udf.register(...) instead.
+```
+
+We decided not to use "SparkSession" in a first step and to follow the error instructions. 
+
+So, we replaced our direct call to Hive by a UDF registration.
+
+## e) Depreciations
+
+We noticed some depreciations like HiveContext or Accumulators for instance. But we decided not to deal with them for the moment.
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: depreciations => deprecations

File: _posts/2017-12-13-spark-2.md [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+## d) Performance
+We have made some gains in performance. 
+Before, running our jobs lasted around three hours. Now, it lasts around two and a half hours.
+
+We hope we will make some other gains by migrating to the Spark 2.2 philosophy ("SparkSession", etc).
+
+## Conclusion
+As there are many backward compatibilities with Spark 2, it is not so difficult to make a first migration to make your project work. Nonetheless, it could be long to validate. It depends on your tests stategy too.
+
+We can add that we did not meet particular problems with the fact that we migrated from Spark 1.5 instead of Spark 1.6.
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: We can add that we did not encounter particular problems while migrating from Spark 1.5 to Spark 1.6.

File: _posts/2017-12-13-spark-2.md [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+
+Our next step now is to integrate the new philosophy of Spark 2.2 to get the best of the new version.
```
Comment (Osirisxxl) [inline _posts/2017-12-13-spark-2.md]: Our next step will be to integrate the new philosophy of Spark 2.2 to get the best of this new version.

File: _posts/2017-12-13-spark-2.md (hunk 5 above)
Comment (NastasiaSaby) [reply _posts/2017-12-13-spark-2.md]: Done ;)

**Comment by CoryChaplin** [comment]
👍

**Pull request merged**